### ISSUE TITLE
GNU parallel throws "Array index out of bounds" errors when abbreviations are set in fish config

### ISSUE THREAD

**Issue by paradox460** [fish-shell/fish-shell#2664]
## Reproduction

Add abbreviations, such as these, to your fish config file:

```fish
abbr b="bundle"
```

The command/abbreviation don't actually seem to matter, as even simply setting them to echo causes the issue.

Run a simple parallel command, i.e. `seq 1 1000 | parallel --gnu -n 1 echo`

You get this output:

```
4
5
8
Array index out of bounds
/usr/local/Cellar/fish/HEAD/share/fish/functions/abbr.fish (line 12): 			__fish_abbr_parse_entry $fish_user_abbreviations[$i] key_i
                                                                      			                                                  ^
in function '__fish_abbr_get_by_key'
	called on line 2 of file /usr/local/Cellar/fish/HEAD/share/fish/functions/abbr.fish
	with parameter list 'bx'

in command substitution
	called on line 0 of file /usr/local/Cellar/fish/HEAD/share/fish/functions/abbr.fish

in function 'abbr'
	called on line 108 of file ~/.config/fish/config.fish
	with parameter list 'bx=bundle exec'

from sourcing file ~/.config/fish/config.fish
	called during startup

1
```

**Comment by faho** [comment]
I can't reproduce this. What OS are you on (OSX, right?), which version of fish?

Can you post your config.fish (and related files) or at least all of your abbrs (or how many exactly)?

The "Array index out of bounds" seems to originate from this:

```fish
		for i in (seq $count)
			set -l key_i
			__fish_abbr_parse_entry $fish_user_abbreviations[$i] key_i
                        # SNIP
		end
```

which seems to suggest that fish_user_abbreviations changed (at least enough entries were removed to make it smaller than $i elements) _after_ the `count` was run.

Either we have a fairly annoying buffering/threading bug, or you've got some event handler or something that does something to fish_user_abbreviations.

**Comment by paradox460** [comment]
I'm on OS X 10.11.2, and on `fish, version 2.2.0-251-g11c8a47`, installed via homebrew with the `--HEAD` option

[my config.fish](https://github.com/paradox460/.dotfiles/blob/master/.config/fish/config.fish)

**Comment by ridiculousfish** [comment]
Thanks. The problem is this stuff in config.fish:

```
abbr b="bundle"
abbr bx="bundle exec"
abbr rk="rake"
abbr rb="ruby"
```

`abbr` is persistent, so putting it in config.fish like this is strictly speaking unnecessary. What this ends up doing is deleting the existing abbreviation and then re-adding it.

Meanwhile in another shell, `__fish_abbr_get_by_key` gets the count of the abbreviations, and then loops through indexes through that count. Since the list of abbreviations is changing, this ends up racing and you can get an array out of bounds error.

Putting `abbr` in config.fish is unnecessary but may be nice to have (e.g. for syncing). So probably the first fix is to make `abbr` a no-op if it's not changing anything, which will improve performance. A second nice-to-have is to make `abbr` robust against concurrent modifications, which means snapshotting `$fish_user_abbreviations` and then writing it back afterwards.

**Comment by paradox460** [comment]
Ah. I suppose a temporary bandaid would be adding a test/guard clause around my abbr stuff, that checks `$fish_user_abbreviations` and doesn't set them if its set/matches the proper value.

**Comment by ridiculousfish** [comment]
Yep, that will unblock you. More generally, if you're invoking fish for scripting, you can wrap the interactive configuration (key bindings, colors, etc.) in `if status --is-interactive`. That will speed up launch for non-interactive scripts.

**Comment by faho** [comment]
Okay, so parallel calls fish, as the user's configured shell. This then reads config.fish, which will re-do a bunch of abbr calls.

The good news is that, in case the abbrs already exist, it won't actually get that far anymore since 585bdf45c8abf6c837ca1e0dd70f8c54beb1833d. Before it would find the index, remove that, and readd it. Right after the removal, the length would obviously be smaller than before, so a `count` before the removal could lead to an invalid index. Now, it'll see it already exists and return early, without modifying the variable.

So this particular error can't happen anymore.

Is there any other case we should fix?

**Comment by faho** [comment]
This only concerns cases where we touch universal variables as part of startup in a non-interactive shell. The only places we `set -U` variables in share/* seem to be:

- The Classic + VCS prompt
- __fish_config_interactive
- abbr

Most of these only _set_ the variables to a default value if they are defined. The only relevant erase I can find is in abbr and that's already mostly fixed.

All these calls have one thing in common: If there's a freak scheduling accident where one fish instance is stopped right after `set -q` while the other goes on to set the variable, it would then reset the variable. But I can't see how that can be exploited to get a different value without _appending_.

In abbr's case, there's still a possible race, though I've not been able to reproduce it - even by writing a script that removes the abbrs and runs parallel (which, btw, has quite an annoying "citation notice" that applies only to academics - apparently they don't consider this use of the software worthy - what's even more annoying is that they'd like a payment of _10000_ € before they absolve you of guilt should you not cite).

In conclusion: This particular issue is fixed and I can't see any more like it. It's still possible that there are others, but I'd suggest opening new issues for those. Thanks!
